registry/storage/driver/filesystem/driver.go

```diff
@@ -291,6 +292,54 @@ func (d *driver) Move(ctx context.Context, sourcePath string, destPath string) e
 	}
 
 	err := os.Rename(source, dest)
+	if err != nil {
+		// Copy then delete fall back
+		linkErr := err.(*os.LinkError)
+		if errno, ok := linkErr.Err.(syscall.Errno); ok && errno == syscall.EXDEV || os.IsExist(err) {
+			err = d.copyFile(source, dest)
+			// If we just moved an uploaded blob, the PurgeUploads loop probably removed it already,
+			// therefore we must check if the source file still exists before trying to delete it
```

Comment:
The extra stat is still not be necessary here. The conclusion should be to use `RemoveAll` instead of `Remove` since `RemoveAll` works fine if the file is already removed. The time between checking the stat and calling `RemoveAll` may also have seen the file removed, there is no extra cost in just always using `RemoveAll`.